### PR TITLE
[ROCm][DSv4.1][Perf] Fused router gate for gfx950

### PR DESCRIPTION
**Pull request by Fangzhou-Ai**
## Purpose and measured status

Optimize the gfx950 BF16 router `X[M,7168] @ W[384,7168].T` plus sqrtsoftplus scoring, correction bias, and expert selection. The tested M values are all **33 distinct products** of `[1,2,4,6,8,16,32,64,128,256] × [1,2,3,4,5,6]`, through 1536; both top-k 6 and 8 are measured below.

The optimized operation beats the runnable unfused baseline at every measured size: **1.47–3.86× with cold rotating inputs**. **It does not reach the raw hardware roofline.** At M=1536/top-k6, the full operation takes **23.17 µs cold** versus **3.450 µs** for the ideal bandwidth/compute-only bound (**14.9%**). The analysis below identifies the remaining costs.

The original draft fails compilation at `tl.arange(0,384)`. The runnable baseline here is BF16→FP32 `torch.mm` (the gate GEMM operation used by GateLinear) followed by `fused_topk_bias(scoring_func="sqrtsoftplus")`.

## Implementation

- M=1–2 uses a vector GEMV; other batches use tiled BF16 MFMA with deterministic FP32 split-K partials.
- The second kernel fuses partial reduction, scoring, bias-aware ranking, normalization, and scaling. Larger batches use a single-wave Gluon selector; smaller batches use packed-key Triton top-k. This is a **two-kernel** implementation.
- Ranking keys retain the entire FP32 score and inverse expert ID. Equal computed ranks prefer the lower expert ID; output weights retain the unbiased score.
- Compensated `log1p` preserves small negative-tail scores without the slower general OCML path. A direct maximum handles top-k1, which otherwise fails to compile with the installed Triton `topk(k=1)`.
- M>768 uses larger MFMA32 tiles and an XCD-aware launch order. The supported token cap increases to 1536; input storage/device and ID dtype validation are enforced.

No serving/model dispatch is wired in this PR. No model-level speedup or evaluation result is claimed.

## Measurement protocol

Measured commit: `f894dba7781cda88aab4edb1fbc8012fc4c2b87f`. MI355X/gfx950, one GPU (physical GPU4), BF16 inputs, FP32 accumulation/output weights, int32 expert IDs, correction bias, renormalization, scaling=1.5. PyTorch `2.12.0+rocm10.0.0`, HIP `7.15.26333`, Triton `3.8.0`. Environment: `HIP_VISIBLE_DEVICES=4 HIP_FORCE_DEV_KERNARG=1 ROCPROFILER_QUEUE_INTERPOSITION=0`.

Both arms measure the complete GPU operation using HIP graph replay. Compilation, allocation, random inputs, and copies are outside timing. Events surround 128 operations; three rounds alternate arm order with 15 samples each. Tables report medians; the JSON artifact retains p10/p90 and kernel source SHA256 `de8dd90626b18179a16bf49b0d3ff8298401f002901da3c71903a368368f0e99`. Correctness is checked before timing.

**Warm** reuses one input set. **Cold** rotates 128 independent input sets: gate weights alone occupy **672 MiB**, exceeding the 256 MB last-level cache. This defeats reuse between operations while preserving legitimate reuse inside each operation. Warm throughput is not evidence of HBM bandwidth. CUPTI is unavailable on ROCm; per-dispatch event timing was rejected because it inflated a no-op control from 1.684 to 8.52 µs.

## Full performance sweep

All latencies are µs. Cold speedup is unfused/optimized. “Roof %” is the ideal operation-level HBM/compute bound divided by optimized cold latency; it excludes launch and selection costs and is **not a hardware utilization counter**.

### Top-k 6

| M | Unfused warm | Optimized warm | Unfused cold | Optimized cold | Cold speedup | Roof % |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 19.11 | 5.10 | 22.06 | 6.13 | 3.60× | 11.3% |
| 2 | 19.54 | 5.47 | 22.67 | 6.55 | 3.46× | 10.6% |
| 3 | 18.39 | 6.47 | 21.97 | 8.71 | 2.52× | 8.0% |
| 4 | 18.42 | 6.55 | 22.22 | 8.80 | 2.52× | 7.9% |
| 5 | 18.54 | 6.49 | 22.27 | 8.81 | 2.53× | 7.9% |
| 6 | 18.81 | 6.53 | 22.39 | 8.88 | 2.52× | 7.9% |
| 8 | 18.89 | 6.55 | 22.64 | 8.96 | 2.53× | 7.8% |
| 10 | 17.34 | 6.58 | 20.63 | 8.98 | 2.30× | 7.9% |
| 12 | 17.04 | 6.62 | 20.89 | 9.00 | 2.32× | 7.9% |
| 16 | 17.35 | 6.67 | 21.32 | 9.05 | 2.35× | 7.9% |
| 18 | 16.12 | 6.73 | 20.40 | 9.25 | 2.21× | 7.8% |
| 20 | 16.09 | 6.75 | 20.42 | 9.30 | 2.20× | 7.8% |
| 24 | 17.59 | 6.84 | 20.92 | 9.40 | 2.22× | 7.8% |
| 30 | 17.67 | 6.90 | 21.39 | 9.56 | 2.24× | 7.8% |
| 32 | 17.51 | 6.92 | 21.31 | 9.48 | 2.25× | 7.9% |
| 36 | 19.09 | 7.02 | 22.73 | 9.56 | 2.38× | 7.9% |
| 40 | 19.16 | 7.06 | 22.83 | 9.67 | 2.36× | 7.9% |
| 48 | 19.27 | 7.08 | 22.95 | 9.72 | 2.36× | 8.0% |
| 64 | 18.63 | 7.34 | 23.66 | 9.91 | 2.39× | 8.1% |
| 80 | 18.41 | 9.47 | 23.79 | 12.45 | 1.91× | 6.7% |
| 96 | 17.60 | 9.57 | 21.73 | 12.52 | 1.74× | 6.9% |
| 128 | 17.56 | 8.21 | 21.71 | 12.04 | 1.80× | 7.6% |
| 160 | 17.62 | 10.73 | 21.85 | 13.28 | 1.65× | 7.4% |
| 192 | 20.82 | 11.22 | 24.40 | 13.67 | 1.79× | 7.6% |
| 256 | 20.23 | 11.00 | 24.87 | 13.90 | 1.79× | 8.3% |
| 320 | 21.31 | 13.48 | 25.63 | 14.49 | 1.77× | 8.7% |
| 384 | 24.09 | 14.20 | 25.67 | 14.99 | 1.71× | 9.2% |
| 512 | 23.26 | 12.72 | 27.69 | 14.48 | 1.91× | 11.1% |
| 640 | 25.24 | 18.35 | 29.56 | 18.75 | 1.58× | 9.8% |
| 768 | 27.39 | 16.77 | 31.23 | 17.56 | 1.78× | 11.8% |
| 1024 | 28.71 | 17.48 | 32.06 | 21.82 | 1.47× | 11.6% |
| 1280 | 32.19 | 18.34 | 36.01 | 22.54 | 1.60× | 13.3% |
| 1536 | 31.35 | 19.40 | 35.07 | 23.17 | 1.51× | 14.9% |

### Top-k 8

| M | Unfused warm | Optimized warm | Unfused cold | Optimized cold | Cold speedup | Roof % |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 21.01 | 4.99 | 23.81 | 6.17 | 3.86× | 11.2% |
| 2 | 21.33 | 5.37 | 24.46 | 6.59 | 3.71× | 10.5% |
| 3 | 20.29 | 6.49 | 23.78 | 8.77 | 2.71× | 7.9% |
| 4 | 20.36 | 6.57 | 24.09 | 8.85 | 2.72× | 7.9% |
| 5 | 20.44 | 6.52 | 24.16 | 8.86 | 2.73× | 7.9% |
| 6 | 20.61 | 6.57 | 24.27 | 8.90 | 2.73× | 7.9% |
| 8 | 20.70 | 6.57 | 24.47 | 8.97 | 2.73× | 7.8% |
| 10 | 18.95 | 6.63 | 22.49 | 9.00 | 2.50× | 7.8% |
| 12 | 18.90 | 6.62 | 22.64 | 9.02 | 2.51× | 7.9% |
| 16 | 19.01 | 6.70 | 22.95 | 9.12 | 2.52× | 7.9% |
| 18 | 17.79 | 6.74 | 21.87 | 9.22 | 2.37× | 7.8% |
| 20 | 17.88 | 6.81 | 22.16 | 9.33 | 2.38× | 7.8% |
| 24 | 19.43 | 6.86 | 22.56 | 9.44 | 2.39× | 7.8% |
| 30 | 19.30 | 6.91 | 23.01 | 9.59 | 2.40× | 7.7% |
| 32 | 19.29 | 6.91 | 23.08 | 9.52 | 2.42× | 7.8% |
| 36 | 20.91 | 7.04 | 24.44 | 9.61 | 2.54× | 7.8% |
| 40 | 20.96 | 7.10 | 24.52 | 9.66 | 2.54× | 7.9% |
| 48 | 21.01 | 7.09 | 24.60 | 9.71 | 2.53× | 8.0% |
| 64 | 20.34 | 7.33 | 25.36 | 9.91 | 2.56× | 8.1% |
| 80 | 20.18 | 9.51 | 25.46 | 12.43 | 2.05× | 6.7% |
| 96 | 19.25 | 9.54 | 23.37 | 12.54 | 1.86× | 6.9% |
| 128 | 19.46 | 8.52 | 23.41 | 12.38 | 1.89× | 7.4% |
| 160 | 19.35 | 11.09 | 23.46 | 13.66 | 1.72× | 7.1% |
| 192 | 23.07 | 11.62 | 26.60 | 14.05 | 1.89× | 7.4% |
| 256 | 21.87 | 11.36 | 26.49 | 14.26 | 1.86× | 8.1% |
| 320 | 23.39 | 13.94 | 27.37 | 14.98 | 1.83× | 8.4% |
| 384 | 26.18 | 14.66 | 27.74 | 15.47 | 1.79× | 8.9% |
| 512 | 25.39 | 13.19 | 29.44 | 14.94 | 1.97× | 10.8% |
| 640 | 27.19 | 18.77 | 31.44 | 19.16 | 1.64× | 9.6% |
| 768 | 28.96 | 17.16 | 33.15 | 17.95 | 1.85× | 11.5% |
| 1024 | 30.64 | 17.89 | 34.09 | 22.24 | 1.53× | 11.4% |
| 1280 | 34.32 | 18.88 | 38.17 | 23.06 | 1.66× | 13.0% |
| 1536 | 33.67 | 20.02 | 37.35 | 23.73 | 1.57× | 14.5% |

## Roofline analysis

The dense MI355X specifications are **8 TB/s HBM**, **2.5166 PFLOP/s BF16 matrix compute**, and **256 MB LLC**. The higher sparse compute number is not applicable. Source: [AMD MI355X datasheet](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/product-briefs/amd-instinct-mi355x-gpu-brochure.pdf).

For N=384, K=7168, and k selected experts:

```text
F = 2*M*N*K
B_min = 2*(M*K + N*K) + 4*N + 8*M*k
AI = F / B_min
T_peak = max(B_min / 8e12, F / 2.5166e15)
TFLOP/s = F / latency_us / 1e6
Effective GB/s = B_min / latency_us / 1000
```

Bytes count each BF16 input and FP32 bias once and one FP32 weight plus int32 ID per selected expert. They exclude scratch and repeated tile loads. The FLOP count includes GEMM, but not nonlinear scoring, comparison, or reduction. Effective bandwidth is mathematical operation traffic, not measured HBM traffic.

Independent large-workload controls on GPU6 reached **6.533 TB/s** for a checked 1 GiB streaming read and **1.403 PFLOP/s** for 8192-square BF16 GEMM with FP32 output after sustained warmup. A live sample during GEMM recorded 1766 MHz SCLK/1399 W, versus the datasheet's boost conditions. These are measured calibration points, not proven ceilings for this router. Their illustrative bound is `max(B_min/6.533e12, F/1.403e15)`.

Representative top-k8 bounds and achieved full-operation throughput:

| M | AI, FLOP/B | Minimum MB | Peak bound, µs | Calibration bound, µs | Cold TFLOP/s | Effective cold GB/s |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 1.00 | 5.521 | 0.690 | 0.845 | 0.89 | 894.2 |
| 16 | 15.35 | 5.737 | 0.717 | 0.878 | 9.65 | 628.8 |
| 128 | 95.87 | 7.350 | 0.919 | 1.125 | 56.90 | 593.5 |
| 256 | 153.30 | 9.193 | 1.149 | 1.407 | 98.84 | 644.8 |
| 512 | 218.84 | 12.879 | 1.610 | 2.009 | 188.63 | 862.0 |
| 768 | 255.22 | 16.566 | 2.071 | 3.013 | 235.60 | 923.1 |
| 1024 | 278.35 | 20.252 | 2.532 | 4.018 | 253.45 | 910.5 |
| 1536 | 306.09 | 27.625 | 3.453 | 6.027 | 356.26 | 1163.9 |

The peak ridge is 314.575 FLOP/B, so the entire requested sweep is memory-bound in that ideal model (M1536/top-k8 reaches 306.09 FLOP/B). The calibration ridge is 214.76 FLOP/B, making its larger-M bound compute-dominated. Neither model captures launch/selection latency or limited workgroup residency.

**Small M:** a graph no-op measured 1.684 µs and a rotating ~5.5 MB read measured about 2.78 µs, already above the 0.69 µs bulk-bandwidth ideal. Two dependent launches, scalar/vector scoring, and wave communication are substantial fixed costs. These controls provide context; subtracting their latency from the operation would not be a valid measured speedup.

**Large M:** split-K trades extra traffic for occupancy. At M1536/S7, partial logits occupy 16.515 MB and add **33.030 MB of logical write+read traffic**. Final BM128/BN128 tiling issues approximately **132.121 MB of logical input loads**, despite only 27.525 MB of unique BF16 input data. Those repeated loads can hit cache; they are not an HBM-byte count.

Counter ablation at M1536, prior BM64/BN128/MFMA16 versus final BM128/BN128/MFMA32/XCD order:

| Counter/resource | Prior tile | Final tile |
| --- | ---: | ---: |
| FetchSize beyond L2, KiB | 64,534.5 | 33,589.5 |
| WriteSize, KiB | 16,128 | 16,128 |
| L2 hit rate | 64.4% | 74.6% |
| MFMA utilization | 13.45% | 15.12% |
| VMEM latency, cycles | 807 | 710 |
| VALU instructions | 1,316,448 | 695,584 |
| Register spills / LDS bank conflicts | 0 / 0 | 0 / 0 |

FetchSize is traffic beyond L2, not proof of off-chip HBM bytes on this LLC-equipped GPU. Counter collection perturbs cache/timing, so these counters are separate from the unprofiled latency tables. The ablation changes tiling, MFMA shape, and ordering together; it does not attribute the entire gain to ordering alone.

The final large tile uses 99,808 B (97.5 KiB) LDS and launches 252 useful CTAs at M1536, allowing only one resident CTA per CU. Lower memory latency and higher cache hit rate support improved locality; low MFMA utilization and limited residency remain consistent with a memory/dependency-latency bottleneck. **The raw roofline gap remains; it is not explained by missing matrix instructions or register spills.**

Further tests rejected no-split GEMM, a fully fused all-expert tile, explicit Gluon double buffering, prepared weights, wider tiles, and increased-residency configurations. They either slowed GEMM or increased partial-buffer/selection cost enough to lose end-to-end. Prepared weights additionally required a ~2.7 µs packing pass and 5.5 MB persistent storage. The final choice preserves the original contiguous weight input format.

## Validation and limits

```bash
HIP_VISIBLE_DEVICES=7 .venv/bin/python -m pytest \
  tests/kernels/moe/test_topk_softplus_sqrt.py -k rocm_fused_router_gate -q
# 53 passed

HIP_VISIBLE_DEVICES=4 .venv/bin/python -m \
  benchmarks.kernels.benchmark_rocm_fused_router_gate \
  --samples 15 --output final_benchmark.json
# 33 M values × 2 top-k values × 2 cache modes = 132 measurements
# Reference checks passed before timing

.venv/bin/pre-commit run --files \
  vllm/model_executor/layers/fused_moe/router/rocm_fused_router_gate.py \
  tests/kernels/moe/test_topk_softplus_sqrt.py \
  benchmarks/kernels/benchmark_rocm_fused_router_gate.py
# Passed
```

The full benchmark agrees with FP32 Torch reference expert IDs exactly and weights at atol=rtol=2e-5. Tests cover masked rows, empty inputs, bias/no bias, both normalization modes, int32/int64 IDs, ties, negative tails, invalid storage, and top-k1.

Additional selector checks passed 240 cases and a 300,001-logit scoring check (maximum relative error 1.95e-6 for logits≥-87). A separate 32-case generic top-k/tail probe preserved all selected expert sets and weights; four cases permuted nearly tied selected experts (reference score gaps≤4.17e-7). Auditing the computed ranks confirmed correct ordering for the kernel's own FP32 scores. This is not a claim of bitwise or batch-invariant equivalence across accumulation implementations. Extreme exponential subnormals can flush on this ROCm configuration.

**Model evaluation:** not run. There is no model callsite dispatching this kernel in this change, and the local checkpoint is inaccessible. Existing evaluation artifacts belong to another change and are not reused as evidence. Serving integration and an evaluation that actually exercises this kernel remain prerequisites for a model-level claim.

## Duplicate work and AI assistance

This updates the existing #57478 rather than opening a competing PR. The open-PR search for `fused router gate` identified this existing BF16 gate/selection work; the optimization is applied here. NVIDIA Mega-Gate work and selection-only ROCm changes do not provide this gfx950 BF16 projection implementation.

AI assistance was used (Codex, including parallel tuning/review agents; the original draft also disclosed Cursor). The human submitter must review the changed lines, understand the tradeoffs, and run the relevant tests before submitting this as ready.

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
I am going open this just to test CI

**Comment by Fangzhou-Ai** [comment]
hi @shen-shanshan  please take a look at this new fused op, it should help both dsv4 and dsv4.1

**Review by shen-shanshan — COMMENTED** [review]
Overall LGTM, since I'm not a kernel expert. Only one question: `rocm_fused_router_gate` is not used by any models, so this PR only added the new fused kernel but not call it?
